aten/src/ATen/native/LossNLL.cpp

```diff
@@ -455,19 +455,74 @@ Tensor nll_loss_backward_cpu(
   return grad_input;
 }
 
+Tensor cross_entropy_loss_prob_target(
+    const Tensor& self,
+    const Tensor& target,
+    const Tensor& weight,
+    int64_t reduction) {
+  const auto n_classes = self.size(1);
+  TORCH_CHECK(
+      !weight.defined() || (weight.dim() == 1 && weight.numel() == n_classes),
+      "weight tensor should be defined either for all ",
+      n_classes,
+      " classes or no classes"
+      " but got weight tensor of shape: ",
+      weight.sizes());
+
+  Tensor ret;
+  Tensor input = at::log_softmax(
+      self, 1, optTypeMetaToScalarType(self.options().dtype_opt()));
+  if (weight.defined()) {
+    // Expand weight to the correct number of dims for broadcasting with input / target
+    auto weight_bc_shape = std::vector<int64_t>(input.dim(), 1);
+    weight_bc_shape[1] = weight.size(0);
+    Tensor weight_ = weight.view(weight_bc_shape);
+
+    ret = -(input * target * weight_).sum(1);
+    if (reduction == Reduction::Mean) {
+      // Compute weighted mean
+      ret = ret.sum() / (target * weight_).sum();
```

Comment:
This... seems like a good argument for why a user would expect reduction=mean to return (-(input * target * weight_).mean()), I'm having a hard time coming up with a use case where someone wants probabilities and wants to do a weighted mean over the probabilities and weights.

At any rate we should probably be consistent with our hard targets cross_entropy function...

Reply:
Yeah, I agree it's super weird, but I did it this way to be consistent with the hard target cross-entropy loss. With one-hot targets, the results are equivalent between soft and hard if done like this :/